utils/posts.js

```diff
@@ -41,15 +42,56 @@ export function getPostMetadata(id) {
             ...matterResult.data,
         },
     };
-
+    
     const title = post.meta.title;
-    var seoTitle = title.toLowerCase().replace(/[^a-z0-9]+/g, ' ').trim().replace(/ /g, '-');
+    const seoTitle = title.toLowerCase().replace(/[^a-z0-9]+/g, ' ').trim().replace(/ /g, '-');
 
     post.meta.seoTitle = seoTitle;
-
+    
     return post;
 }
 
+// Finds all the unique tags in the tag cloud and formats them for the dynamic route
+export function getUniqueTags() {
+    const posts = getPosts();
+    var uniqueTags = [];
+    
+    for(var i = 0; i < posts.length; i++) {
+        if(posts[i].meta.tags) {
+            const  tags = posts[i].meta.tags.split(', ');
+            for(var j = 0; j < tags.length; j++) {
+                uniqueTags.push(tags[j].replace(/ /g, '-'));
+            }
+        };
+    };
+    
+    uniqueTags = [...new Set(uniqueTags)];
+    uniqueTags.sort(function (a, b) {
+	return a.toLowerCase() < b.toLowerCase() ? -1 : 1;
+    });
+    
+    return uniqueTags;
+}
+
+// Returns all posts sorted into their respective tag(s)
+export function getTagSortedPosts() {
+    const posts = getPosts();
+    const tags = getUniqueTags();
+    var sortedPosts = {};
+    for(var i = 0; i < tags.length; i++) {
+	sortedPosts[tags[i]] = [];
+    };
+    for(var i = 0; i < posts.length; i++) {
+        if(posts[i].meta.tags) {
```

Comment:
This is the second time this method is implemented. We should define a method for getting the tags in a more central location.

Perhaps we should just do it inside of `getPostMetadata`, and there, convert `post.meta.tags` to `post.tags` which is a list/array containing values that are lowercased and trimmed.

Reply:
@jontsai I think I will make a constant in `utils/posts.js` called `convertTags` that basically is a series of key-value pairs so I can easily convert the tags from the form used in the dynamic route to the form used in the page content. I think this will simplify the task and make it easier to code. Are you ok with that?

Reply:
@tang0226 that is a step in the right direction, yes, but I believe that is not the final form.

Please keep the **principle** DRY (https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) in mind, but the actual implementation will take on many different forms. It's hard to explain my proposal short of writing the code out myself, so I suggest you take a stab at DRY-ing up the code yourself first, and then see if you can shorten it as much as possible, and utilize OOP patterns (https://en.wikipedia.org/wiki/Object-oriented_programming)